profiles/serializers_test.py

```diff
@@ -304,3 +309,158 @@ def test_delete_employment(self):
         assert employment3.profile.work_history.count() == 1
+
+
+class ProfileFilledOutTests(ESTestCase):
+    """Tests for validating filled out profiles"""
+
+    def create_profile(self):
+        """
+        Create a profile and social auth
+        """
+        with mute_signals(post_save):
+            profile = ProfileFactory.create(
+                filled_out=True,
+                agreed_to_terms_of_service=True,
+            )
+            EmploymentFactory.create(profile=profile)
+            EducationFactory.create(profile=profile)
+            profile.user.social_auth.create(
+                provider=EdxOrgOAuth2.name,
+                uid="{}_edx".format(profile.user.username)
+            )
+            return profile
+
+    def test_success(self):
+        """
+        Test a successful validation
+        """
+        profile = self.create_profile()
+
+        serializer = ProfileFilledOutSerializer(profile)
+        data = serializer.to_representation(profile)
+        ProfileFilledOutSerializer(data=data).is_valid(raise_exception=True)
+
+    def test_filled_out_false(self):
+        """
+        filled_out cannot be set to false
+        """
+        profile = self.create_profile()
+
+        serializer = ProfileFilledOutSerializer(profile)
+        data = serializer.to_representation(profile)
+        data['filled_out'] = False
+        with self.assertRaises(ValidationError) as ex:
+            ProfileFilledOutSerializer(data=data).is_valid(raise_exception=True)
+        assert ex.exception.args[0] == {'non_field_errors': ['filled_out cannot be set to false']}
+
+    def test_agreed_to_terms_of_service_false(self):
+        """
+        agreed_to_terms_of_service cannot be set to false
+        """
+        profile = self.create_profile()
+
+        serializer = ProfileFilledOutSerializer(profile)
+        data = serializer.to_representation(profile)
+        data['agreed_to_terms_of_service'] = False
+        with self.assertRaises(ValidationError) as ex:
+            ProfileFilledOutSerializer(data=data).is_valid(raise_exception=True)
+        assert ex.exception.args[0] == {'non_field_errors': ['agreed_to_terms_of_service cannot be set to false']}
+
+    def test_required_fields(self):
+        """
+        All fields are required after a profile is marked filled_out
+        """
+        profile = self.create_profile()
+
+        serializer = ProfileFilledOutSerializer(profile)
+        data = serializer.to_representation(profile)
+        # skip fields that are generated, and work_history and education which are tested elsewhere
+        keys = set(ProfileFilledOutSerializer.Meta.fields) - {
+            'username',
+            'profile_url_full',
+            'profile_url_large',
+            'profile_url_medium',
+            'profile_url_small',
+            'pretty_printed_student_id',
+            'work_history',
+            'education',
+        }
+        for key in keys:
```

Comment:
whipped this together:

```python
class ReqProfileFieldTester:
    def __init__(self, test_obj):
        self.test = test_obj
        self.clone = {}

    def expected_req_field_exception(self, field_key, message):
        return {field_key: [message]}

    def set_value_at_key(self, field_key, value):
        self.clone[field_key] = value

    def get_field_from_key(self, field_key):
        return ProfileFilledOutSerializer(data=self.clone).fields[field_key]

    def check_required_fields(self, keys, data):
        for key in keys:
            self.clone = deepcopy(data)
            self.set_value_at_key(key, None)
            with self.test.assertRaises(ValidationError) as ex:
                ProfileFilledOutSerializer(data=self.clone).is_valid(raise_exception=True)
            assert ex.exception.args[0] == self.expected_req_field_exception(
                key, 'This field may not be null.'
            )

            field = self.get_field_from_key(key)
            if isinstance(field, CharField):
                # test blank string too
                self.set_value_at_key(key, '')
                with self.test.assertRaises(ValidationError) as ex:
                    ProfileFilledOutSerializer(data=self.clone).is_valid(raise_exception=True)
                assert ex.exception.args[0] == self.expected_req_field_exception(
                    key, 'This field may not be blank.'
                )


class ReqProfileNestedFieldTester(ReqProfileFieldTester):
    def __init__(self, test_obj, nested_key=''):
        self.nested_key = nested_key
        super(ReqProfileNestedFieldTester, self).__init__(test_obj)

    def expected_req_field_exception(self, field_key, message):
        return {self.nested_key: [{field_key: [message]}]}

    def set_value_at_key(self, field_key, value):
        self.clone[self.nested_key][0][field_key] = value

    def get_field_from_key(self, field_key):
        return ProfileFilledOutSerializer(data=self.clone).fields[self.nested_key].child.fields[field_key]
```

invocation changes to:

```python
req_field_tester = ReqProfileFieldTester(self)
req_field_tester.check_required_fields(keys, self.data)
# or...
req_field_tester = ReqProfileNestedFieldTester(self, nested_key='work_history')
req_field_tester.check_required_fields(keys, self.data)
```

i'm sort of on the line as to whether or not this is overkill. all i know is that this functionality (iterating through fields, asserting the right validation errors with a none value or blank value for character fields) is general enough and complex enough that it's probably worth some abstraction and not preferable to repeat it 3 times over